Networking/WebSockets/WebSocketImpl.cc

```diff
@@ -454,6 +454,18 @@ namespace litecore { namespace websocket {
     // Called when the underlying socket closes.
     void WebSocketImpl::onClose(CloseStatus status) {
         {
+            if (!hasDelegate()) {
```

Comment:
This conditional does not match the description in the commit, and requires comment analysis.  I feel like this should be `if(role() != Role::Client)` which more clearly identifies the information we are after.

Reply:
I use this as a test because I found server role may also take a factory as argument, c.f.
C4Socket* C4Socket::fromNative(const C4SocketFactory &factory,
                               void *nativeHandle,
                               const C4Address &address)
{
    // Note: This should be wrapped in `retain()` since `C4SocketImpl` is ref-counted,
    // but doing so would cause client code to leak. Instead I added a warning to the doc-comment.
    return new C4SocketImpl(address.toURL(), Role::Server, {}, &factory, nativeHandle);
}
This makes me this, "roll() == Role::Server" may also call factory method acting like a client in the dialog of client v server. If it calls factory.open(), then it would get a delegate and expect onX methods.
The common factor is, WebSocket::connect(Delegate*). As far as I can see from the source code, it seems that role being Role::Server cannot bar it from calling factory.open, etc. Otherwise, why is factory is given to C4Socket of server role? I can change the test to Role() != Role::Client if I am sure that in this case, "factory" passed in the constructor shall not be used.

As for why I only put the test onClose, that's because I find it's not easy for the client code, Java in particular, to call other onX methods because it needs to pass the opaque C4Socket* handle which is passed to client by Core as argument of factory::open. Only onClose is little confusing. It is called in, for instance, in Java's C4Socket::release(), which thinks I am closed because I am reaching end of my life. Other onX methods have clear semantics. In other words, I would think it as a program error leading to UB. It seems to be least disruptive now.

Reply:
Server side (passive), by definition, does not open connections from itself.  It sits there and listens for, and accepts incoming connections.  The factory is given to the server side socket so that it can know how to correctly write bytes and acknowledge receives.  Opening and closing connections is the responsibility of the active (client) side.

It may be that `c4socket_closed` is the only dangerous method, but it's worth checking out the path of the others like `c4socket_opened` to see if they lead to any delegate calls.  I don't think whether or not a platform calls them is important, what is important is that they are able to be called without arbitrarily causing a native crash.

Reply:
Okay, I will add a DBAssert to connect(Delegate*) to assert Role() == client. Then we can just check Role() == client in the client callback methods.
There are 6 methods in WebSocketImp that use delegate(). They all need be guarded against misuse. Very ugly. In the future, we probably should do what @bmeike suggested to subclass them into two, client & server.

Reply:
"Server side (passive), by definition, does not open connections from itself. " -- this statement is incorrect. The following LoopbackWebSocket is created with role == server as follows,
        _replServer = new Replicator(dbServer,
                                     new LoopbackWebSocket(alloc_slice("ws://cli/"_sl), Role::Server, kLatency),
                                     *this, optsRef2);
but hits my assertion:
    void WebSocket::connect(Delegate *delegate) {
//        DebugAssert(role() == Role::Client);
        if (role() != Role::Client) {
            printf("hey got you.\n");
        }
We have to think the role as replicator's role, as supposed to the connection's role.

Reply:
In that case there is no choice but in my opinion it's now even messier than you suggested because we shouldn't be thinking about what a higher level component is doing when working in a lower level component....